pkg/controller/elasticsearch/user/reconcile.go

```diff
@@ -0,0 +1,193 @@
+	"reflect"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/record"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/reconciler"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/tracing"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/watches"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/client"
+	esclient "github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/client"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/label"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/user/filerealm"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/maps"
+
+	"go.elastic.co/apm"
+
+	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+)
+
+var log = logf.Log.WithName("elasticsearch-user")
+
+// ReconcileUsersAndRoles fetches all users and roles and aggregates them into a single
+// Kubernetes secret mounted in the Elasticsearch Pods.
+// That secret contains the file realm files (`users` and `users_roles`) and the file roles (`roles.yml`).
+// users are aggregated from various sources:
+// - predefined users include the controller user, the probe user, and the public-facing elastic user
+// - associated users come from resource associations (eg. Kibana or APMServer)
+// - user-provided users from file realms referenced in the Elasticsearch spec
+// Roles are aggregated from:
+// - predefined roles (for the probe user)
+// - user-provided roles referenced in the Elasticsearch spec
+func ReconcileUsersAndRoles(
+	ctx context.Context,
+	c k8s.Client,
+	es esv1.Elasticsearch,
+	watched watches.DynamicWatches,
+	recorder record.EventRecorder,
+) (client.UserAuth, error) {
+	span, _ := apm.StartSpan(ctx, "reconcile_users", tracing.SpanTypeApp)
+	defer span.End()
+
+	// build aggregate roles and file realms
+	roles, err := aggregateRoles(c, es, watched, recorder)
+	if err != nil {
+		return client.UserAuth{}, err
+	}
+	fileRealm, controllerUser, err := aggregateFileRealm(c, es, watched, recorder)
+	if err != nil {
+		return client.UserAuth{}, err
+	}
+
+	// reconcile the aggregate secret
+	if err := reconcileRolesFileRealmSecret(c, es, roles, fileRealm); err != nil {
+		return client.UserAuth{}, err
+	}
+
+	// return the controller user for next reconciliation steps to interact with Elasticsearch
+	return controllerUser, nil
+}
+
+func getExistingFileRealm(c k8s.Client, es esv1.Elasticsearch) (filerealm.Realm, error) {
+	var secret corev1.Secret
+	if err := c.Get(RolesFileRealmSecretKey(es), &secret); err != nil {
+		return filerealm.Realm{}, err
+	}
+	return filerealm.FromSecret(secret)
+}
+
+// aggregateFileRealm aggregates the various file realms into a single one, and returns the controller user auth.
```

Comment:
Do you mean we would rename `esclient.UserAuth` into `esclient.UserCredentials`?

Reply:
I just had to check the implementation of `esclient.UserAuth` to understand what was a "user auth" and I was wondering if there is not a better name.
 `returns the controller user credential` seems more familiar to me when speaking about what is used by a user to prove that it is the one it pretends to be.

It could be `UserCredential` or `UsernamePasswordCredential` to make it clear that we are using a username and a password (I _think_ singular is fine).

No strong opinion though.

Reply:
I think I called it `Auth` because it will be passed on as an `Authorization` header in the client.

Reply:
Would it make sense to name it `esclient.BasicAuth`?
We could also move it to something like `utils/http/basicauth.go`. I don't think there's a generic `BasicAuth` type we can reuse from go stdlib.

Reply:
I renamed the type to `esclient.BasicAuth`, and changed the user pkg function names to better match "credentials".